src/core/ai/llm/messages/utils.ts

```diff
@@ -149,3 +149,88 @@ export function getFileData(filePart: {
     console.warn('Unexpected file data type in getFileData:', typeof data);
     return '';
 }
+
+import { validateFileForLLM } from '../validation.js';
+
+export interface FilteringConfig {
+    provider: string;
+    model?: string;
+}
+
+/**
+ * Filters message content based on LLM capabilities.
+ * Removes unsupported file attachments while preserving supported content.
+ * Uses model-specific validation when available, falls back to provider-level.
+ * @param messages Array of internal messages to filter
+ * @param config The LLM configuration (provider and optional model)
+ * @returns Filtered messages with unsupported content removed
+ */
+export function filterMessagesByLLMCapabilities(
```

Comment:
nice! this is a very useful helper function

let's add tests for this though, it's hard to validate edge cases manually

(combinations of audio, filetype for different models and validate the history that comes back)